<H2>Importing Libraries</H2>

In [1]:
#Import Libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Create path

path = r'C:\Users\supri\Desktop\Supriya\DataAnalysis\Project\Data Immersion\Achievement 1-2\Instacart_Basket_Analysis'

In [3]:
#Import ords_prods_merge dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))

In [4]:
#Create a smaller subset of data -due to the large data frame

df=ords_prods_merge[:1000000]

In [5]:
#Check output

df.shape

(1000000, 17)

<H3>If Statement with user defined function</H3>

In [6]:
#Define a function

def price_label(row):

    if row ['prices']<= 5:
        return 'Low range product'
    elif (row['prices']> 5) and (row['prices']<=15):
        return 'Mid range product'
    elif row['prices'] >15 :
        return 'High range product'
    else:return 'Not enough data'

In [7]:
#Apply the function

df['price_range'] =df.apply(price_label,axis =1)

C:\Users\supri\AppData\Local\Temp\ipykernel_50824\4268074631.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] =df.apply(price_label,axis =1)


In [8]:
#Check output

df['price_range'].value_counts(dropna = False)

price_range
Mid range product     673183
Low range product     314392
High range product     12425
Name: count, dtype: int64

In [9]:
#Most expensive product in the subset

df['prices'].max()

99999.0

<H4>Using the loc.function on my entire dataframe for the previous message</H4>

In [10]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\supri\AppData\Local\Temp\ipykernel_50824\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
#Check
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

<H5>Use loc() method on entire ords_prods_merge dataframe to apply price range function</H5>

In [14]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
#Check value counts for ords_prods_merge
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

<H4>If Statements with For loops</H4>

In [18]:
# creating a test-for-loop
for x in range(30,45):
    print("My age is %d" %(x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [19]:
#frequency of the “orders_day_of_week”

ords_prods_merge['orders_days_of_week'].value_counts(dropna=False)

orders_days_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [20]:
#Creating a 'busiest day' column that shows busiest day, least busy, and regularly busy days.

result =[]

for value in ords_prods_merge["orders_days_of_week"]:
    if value ==0:
        result.append("Busiest Day")
    elif value==4:
        result.append("Least Day")
    else:
        result.append("Regularly busy")

In [21]:
#Cehck output

result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'Least Day',
 'L

In [22]:
# add result to ords_prods_merge as a new column

ords_prods_merge['busiest_day'] = result

In [23]:
ords_prods_merge['busiest_day'].value_counts(dropna =False)

busiest_day
Regularly busy    22416875
Busiest Day        6204182
Least Day          3783802
Name: count, dtype: int64

In [24]:
ords_prods_merge.head()

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0             0   2539329        1    prior             1   
1             0   2539329        1    prior             1   
2             0   2539329        1    prior             1   
3             0   2539329        1    prior             1   
4             0   2539329        1    prior             1   

   orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   

   first_time_customers  product_id  add_to_cart_order  reordered  \
0                  True         196                  1          0   
1                  True       14084                  2          0   
2                  True       12427                  3          0   
3                  True       26088                  4          0   
4                  True       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    price_range_loc     busiest_day  
0  both  Mid-range product  Regularly busy  
1  both  Mid-range product  Regularly busy  
2  both  Low-range product  Regularly busy  
3  both  Low-range product  Regularly busy  
4  both  Low-range product  Regularly busy

<H4>Update "busiest day" column to "busiest days" to identify order day of the week as \"Busiest days,\" \"Slowest days,\" or \"Regularly busy</H4>

In [25]:
# Create column 'busiest_days" using the frequency of "orders_days_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days)

result_2 =[]

for value in ords_prods_merge["orders_days_of_week"]:
    if value ==0 or value==1:
        result_2.append("Busiest Days")
    elif value==4 or value==3:
        result_2.append("Slowest Days")
    else:
        result_2.append("Regularly busy")

In [26]:
result_2

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Day

In [27]:
# add result to ords_prods_merge as a new column
ords_prods_merge['busiest_days'] = result_2

In [28]:
ords_prods_merge['busiest_days'].value_counts(dropna =False)

busiest_days
Regularly busy    12916111
Busiest Days      11864412
Slowest Days       7624336
Name: count, dtype: int64

In [29]:
#Check df
ords_prods_merge.head()

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0             0   2539329        1    prior             1   
1             0   2539329        1    prior             1   
2             0   2539329        1    prior             1   
3             0   2539329        1    prior             1   
4             0   2539329        1    prior             1   

   orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   

   first_time_customers  product_id  add_to_cart_order  reordered  \
0                  True         196                  1          0   
1                  True       14084                  2          0   
2                  True       12427                  3          0   
3                  True       26088                  4          0   
4                  True       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    price_range_loc     busiest_day    busiest_days  
0  both  Mid-range product  Regularly busy  Regularly busy  
1  both  Mid-range product  Regularly busy  Regularly busy  
2  both  Low-range product  Regularly busy  Regularly busy  
3  both  Low-range product  Regularly busy  Regularly busy  
4  both  Low-range product  Regularly busy  Regularly busy

<H5>Observation- The total number of busiest days is the sum of 0 & 1 (11864412) and for slowest days, the sum of 3 & 4 (7624336) ie Saturday and Sunday are the busiest days and Tuesday and Wednesday seem to be the slowest days.</H5>

<H4>When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, periods of time to be labeled Most orders, Average orders, and Fewest orders.Create a new column containing these labels called busiest_period_of_day.</H4>

In [30]:
#Check frequency of the 'order_hour_of_day' column

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [32]:
# Create a new column 'Busiest_period_of_day' with the following three values
# * Most orders (9 - 16) 
# * Average orders (7-8 and 17 - 22)
# * Fewest orders  (23,0- 6)


result3 = []

for value in ords_prods_merge["order_hour_of_day"]:
  if value in [9,10,11,12,13,14,15,16]:
      result3.append("Most orders")
  elif value in [17,7,8,18,19,20,21,22]:
      result3.append("Average orders")
  else:result3.append("Fewest orders")

In [33]:
result3

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [34]:
#Create new column busiest_period_of_day using result3
ords_prods_merge['busiest_period_of_day'] = result3

In [35]:
ords_prods_merge.head()

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0             0   2539329        1    prior             1   
1             0   2539329        1    prior             1   
2             0   2539329        1    prior             1   
3             0   2539329        1    prior             1   
4             0   2539329        1    prior             1   

   orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   

   first_time_customers  product_id  ...  reordered  \
0                  True         196  ...          0   
1                  True       14084  ...          0   
2                  True       12427  ...          0   
3                  True       26088  ...          0   
4                  True       26405  ...          0   

                              product_name aisle_id  department_id  prices  \
0                                     Soda       77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                      Original Beef Jerky       23             19     4.4   
3               Aged White Cheddar Popcorn       23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls       54             17     1.0   

   match    price_range_loc     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy  Regularly busy   
2   both  Low-range product  Regularly busy  Regularly busy   
3   both  Low-range product  Regularly busy  Regularly busy   
4   both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  

[5 rows x 21 columns]

In [36]:
#Frequency of the busiest_period_of_day

ords_prods_merge['busiest_period_of_day'].value_counts(dropna= False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [37]:
#Export to pickle file

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived.pkl'))